### DISCUSSION TITLE
How to remove preference contributions

### DISCUSSION THREAD

**Discussion by arunswarnam** [eclipse-theia/theia#9144] [Q&A] [answered]
Hi! I have an use-case where I would like to restrict opening new workspace's to current browser tab. 
I was able to override [this](https://github.com/eclipse-theia/theia/blob/master/packages/workspace/src/browser/workspace-service.ts#L498) in my extension and achieve this behavior. 

I still see this preference option showing up on the settings UI which I would prefer to get rid of:
![image](https://user-images.githubusercontent.com/10571107/109575840-72dac880-7aa7-11eb-97d5-260fa23fd873.png)

I see a workspace preference contribution [here](https://github.com/eclipse-theia/theia/blob/master/packages/workspace/src/browser/workspace-preferences.ts#L60) but not sure how to override the behavior. Can someone suggest?

**Comment by kittaakos** [comment, marked as answer]
> I was able to override [this](https://github.com/eclipse-theia/theia/blob/master/packages/workspace/src/browser/workspace-service.ts#L498) in my extension and achieve this behavior.

👍 

> I see a workspace preference contribution [here](https://github.com/eclipse-theia/theia/blob/master/packages/workspace/src/browser/workspace-preferences.ts#L60) but not sure how to override the behavior. Can someone suggest?

You cannot do that, unfortunately. We need to change the Theia code first. The workspace preference schema contribution is a constant and you cannot customize it via dependency injection.

I created an example, what we need to change to make the preferences customizable. Ideally, we have to do the same everywhere, for all other preferences. See the example here https://github.com/kittaakos/theia/commit/e2439d6897ea049f3d6305ba1357c17aa91f1966

Most likely, the first-ever constant preference schema contribution was copied all over the app without thinking about how to extend it. If you are willing to bring this change to Theia, I am happy to assist you and review your changes.

Once we can customize the preferences, all you need to do is to rebind the default behavior. See the example here: https://github.com/kittaakos/theia/commit/a9f4e2e1686db1d70c3db323967a7096bb359110

Finally, you can get rid of the undesired preference.
![Screen Shot 2021-03-02 at 11 19 15](https://user-images.githubusercontent.com/1405703/109634090-21831580-7b49-11eb-89e5-fce4c0eb8235.jpg)

I hope this helps.

**Reply by arunswarnam** [reply]
Thanks for the response and the samples @kittaakos . Sure, I will be happy to bring this change to Theia.

**Reply by KR155E** [reply]
Was this ever implemented?

**Reply by vince-fugnitto** [reply]
@KR155E preference schemas can now be more easily rebinded thanks to https://github.com/eclipse-theia/theia/pull/9883.

**Reply by KR155E** [reply]
Thanks, @vince-fugnitto, that sounds promising. However, my understanding of inversify is limited and I can't find out how to actually remove a preference. Could you give an advice?

I tried the following:

`rebind(ElectronWindowPreferences).toDynamicValue(ctx => {
        const preferences = ctx.container.get<PreferenceService>(PreferenceService);
        const contribution = ctx.container.get<PreferenceContribution>(ElectronWindowPreferenceContribution);
        delete contribution.schema.properties['window.zoomLevel'];
        return createElectronWindowPreferences(preferences, contribution.schema);
    }).inSingletonScope();`

**Reply by vince-fugnitto** [reply]
@KR155E here is an example that was originally part of the pull-request for test purposes (it rebinds `WorkspacePreferenceContribution`) to remove all previous preferences: https://github.com/eclipse-theia/theia/commit/7636170c53b7c327b406588d7c226e35773e4c8b

```ts
rebind(WorkspacePreferenceContribution).toConstantValue({
    schema: <PreferenceSchema>{
        type: 'object',
        properties: {}
    }
});
```

**Reply by KR155E** [reply]
Thank you very much, @vince-fugnitto, that did the trick.
